src/Microsoft.Dnx.Compilation.CSharp.Common/MetadataReferenceExtensions.cs

```diff
@@ -0,0 +1,77 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Reflection.PortableExecutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.Dnx.Compilation.Caching;
+
+namespace Microsoft.Dnx.Compilation.CSharp
+{
+    /// <summary>
+    /// Extension methods for <see cref="IMetadataReference"/>
+    /// </summary>
+    public static class MetadataReferenceExtensions
+    {
+        /// <summary>
+        /// Converts an <see cref="IMetadataReference"/> to a <see cref="MetadataReference"/> instance.
+        /// </summary>
+        /// <param name="metadataReference">The <see cref="IMetadataReference"/> to convert.</param>
+        /// <param name="cache">The <see cref="ICache"/>.</param>
+        /// <param name="onAssemblyMetadataCacheMiss">Invoked when a cache miss occurs when attempting to look 
+        /// up cached <see cref="AssemblyMetadata"/>.</param>
+        /// <returns>The converted <see cref="MetadataReference"/>.</returns>
+        public static MetadataReference ConvertMetadataReference(
+            this IMetadataReference metadataReference,
+            ICache cache,
+            Action<CacheContext, string> onAssemblyMetadataCacheMiss)
+        {
+            var roslynReference = metadataReference as IRoslynMetadataReference;
+
+            if (roslynReference != null)
+            {
+                return roslynReference.MetadataReference;
+            }
+
+            var embeddedReference = metadataReference as IMetadataEmbeddedReference;
+
+            if (embeddedReference != null)
+            {
+                return MetadataReference.CreateFromImage(embeddedReference.Contents);
+            }
+
+            var fileMetadataReference = metadataReference as IMetadataFileReference;
+
+            if (fileMetadataReference != null)
+            {
+                var path = fileMetadataReference.Path;
+                var metadata = cache.Get<AssemblyMetadata>(path, ctx =>
+                {
+                    onAssemblyMetadataCacheMiss(ctx, path);
+
+                    using (var stream = File.OpenRead(path))
+                    {
+                        var moduleMetadata = ModuleMetadata.CreateFromStream(stream, PEStreamOptions.PrefetchMetadata);
+                        return AssemblyMetadata.Create(moduleMetadata);
+                    }
+                });
+
+                return metadata.GetReference(filePath: path);
+            }
+
+            var projectReference = metadataReference as IMetadataProjectReference;
+            if (projectReference != null)
+            {
+                using (var ms = new MemoryStream())
+                {
+                    projectReference.EmitReferenceAssembly(ms);
+
+                    return MetadataReference.CreateFromImage(ms.ToArray());
```

Comment:
We create it - https://github.com/aspnet/dnx/pull/2595/files#diff-6d73c10696b867e7a9c5b6a4243462d9R53